File: Daily Challenge/2020/October/21_asteroid_collision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        res = []

        for x in asteroids:
            # does it collide?
            # Inside the while loop there will be a colission
            while len(res) and x < 0 and res[-1] > 0:
                # Both have same maghnitude so they both break
                if (res[-1] + x) == 0:
                    res.pop()
                    break

                # Incoming asteroid breaks
                elif abs(x) < res[-1]:
                    break

                # top asteroid in stack breaks
                else:
                    res.pop()
                    continue
            else:
                # appending asteroid to res
                res.append(x)

        return res

File: Daily Challenge/2020/October/21_asteroid_collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        stack<int> s;
        bool flag = false; //flag for two cases where incoming asteroid is destroyed 
                           //and shouldn't be pushed
        for(int i = 0; i < asteroids.size(); i++){
            //collision
            flag = false;
            while(!s.empty() and s.top() > 0 and asteroids[i] < 0){
                //both asteroids destroyed
                if(abs(s.top()) == abs(asteroids[i])){
                    s.pop();
                    flag = true;
                    break;
                }
                //incoming asteroid still intact
                else if(abs(s.top()) < abs(asteroids[i]))
                    s.pop();
                //incoming asteroid destroyed
                else if(abs(s.top()) > abs(asteroids[i])){
                    flag = true;
                    break;
                }
                    
            }
            //no collision
            if(!flag)
                s.push(asteroids[i]);
        }
        vector<int> ans;
        while(!s.empty()){
            ans.push_back(s.top());
            s.pop();
        }
        reverse(ans.begin(), ans.end());
        return ans;
    }
};